### PR TITLE
Add tests

### PR DESCRIPTION
**Pull request by j201**
Written by @Barry0121

### CODE REVIEW

File: tests/test_cls.py [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+V1_CLS_ACC = 0.9916
+TOLERANCE = 2e-2
+
+
+class TestClassification(unittest.TestCase):
+
+    @classmethod
+    def setUpClass(cls):
+        """Setup dataset and the v1.2 model, then gather predictions."""
+        seed_everything(42)
```
Comment (anthonycaterini) [inline tests/test_cls.py]: make this a constant top of the file

File: tests/device_utils.py [hunk 2]
```diff
@@ -0,0 +1,34 @@
+"""Shared test helpers: device selection and RNG seeding.
+
+Tests prioritize GPU; CPU is used only when no GPU has enough free memory available.
+"""
+
+
+def seed_everything(seed: int = 42) -> None:
```
Comment (anthonycaterini) [inline tests/device_utils.py:7]: this and the default min GB available should both be constants top of the file

File: tests/test_cls.py [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+        X, y = load_digits(return_X_y=True)
+        cls.X_train, cls.X_test, cls.y_train, cls.y_test = train_test_split(X, y, test_size=0.33, random_state=42)
```
Comment (anthonycaterini) [inline tests/test_cls.py]: 0.33 should be a const also, and `random_state` draw from the same const as `seed_everything`

File: tests/test_cls.py [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+DEVICE = pick_device()
+
+V1_CLS_ACC = 0.9916
+TOLERANCE = 2e-2
+
+
+class TestClassification(unittest.TestCase):
+
+    @classmethod
+    def setUpClass(cls):
```
Comment (anthonycaterini) [inline tests/test_cls.py]: why camelCase?

File: tests/test_cls.py [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+        """Setup dataset and the v1.2 model, then gather predictions."""
+        seed_everything(42)
+        X, y = load_digits(return_X_y=True)
+        cls.X_train, cls.X_test, cls.y_train, cls.y_test = train_test_split(X, y, test_size=0.33, random_state=42)
+        cls.base_model = TabDPTClassifier(device=DEVICE)
```
Comment (anthonycaterini) [inline tests/test_cls.py]: why is this one a const

File: tests/test_cls.py [hunk 6]
```diff
@@ -0,0 +1,51 @@
...
+        cls.base_model.fit(cls.X_train, cls.y_train)
+        cls.y_train_pred = cls.base_model.predict(
+            cls.X_train, n_ensembles=8, temperature=0.8, context_size=None, permute_classes=True, seed=42,
```
Comment (anthonycaterini) [inline tests/test_cls.py]: make these params consts top of the file or the class, here and below, to ensure apples-to-apples comparison between these two lines

File: tests/test_cls.py [hunk 7]
```diff
@@ -0,0 +1,51 @@
...
+        )
+        cls.y_test_pred = cls.base_model.predict(
+            cls.X_test, n_ensembles=8, temperature=0.8, context_size=None, permute_classes=True, seed=42,
+        )
+
+    def test_accuracy(self):
+        """Train prediction should be perfect; v1.2 test accuracy should be at least v1.1's."""
```
Comment (anthonycaterini) [inline tests/test_cls.py:32]: We really expect perfect prediction for training?

File: examples/cls_example.py
Comment (anthonycaterini) [inline examples/cls_example.py:1]: Should update `README.md` to reflect the changes in this path and `reg_example.py`

File: tests/test_cls.py [hunk 8]
```diff
@@ -0,0 +1,51 @@
...
+        train_acc = accuracy_score(self.y_train, self.y_train_pred)
+        test_acc = accuracy_score(self.y_test, self.y_test_pred)
+        print(f"\n[cls] v1.2 test acc {test_acc:.4f} vs v1.1 {V1_CLS_ACC:.4f} (delta {test_acc - V1_CLS_ACC:+.4f})")
+
+        self.assertGreaterEqual(
+            train_acc, 0.999,
```
Comment (anthonycaterini) [inline tests/test_cls.py]: relatedly, this `0.999` should be a const and the error message updated

File: tests/test_estimator.py [hunk 9]
```diff
@@ -0,0 +1,84 @@
...
+from sklearn.preprocessing import (
+    MinMaxScaler,
+    PowerTransformer,
+    QuantileTransformer,
+    RobustScaler,
+    StandardScaler,
+)
+
+from tabdpt import TabDPTClassifier
+from tabdpt import estimator
```
Comment (anthonycaterini) [inline tests/test_estimator.py:13]: do we really need to `import estimator` here?

File: tests/test_inference.py [hunk 10]
```diff
@@ -0,0 +1,149 @@
...
+DEVICE = pick_device()
+N_TRAIN = 60
+
+
+class TestClassifier(unittest.TestCase):
+    @classmethod
+    def setUpClass(cls):
+        seed_everything(0)
+        cls.X = np.random.normal(size=(N_TRAIN, 5)).astype(np.float32)
+        cls.y = np.random.randint(0, 3, N_TRAIN)
```
Comment (anthonycaterini) [inline tests/test_inference.py]: feels like we could have `N_CLASS` defined at the top of this class since it is used a lot

File: tests/test_inference.py [hunk 11]
```diff
@@ -0,0 +1,149 @@
...
+
+    def test_retrieval_path(self):
+        p = self.model.predict_proba(self.X, context_size=20)
+        self.assertEqual(p.shape, (N_TRAIN, 3))
+        np.testing.assert_allclose(p.sum(axis=1), 1.0, atol=1e-4)
+
+    def test_return_logits(self):
+        logits = self.model.predict_proba(self.X, context_size=2048, return_logits=True)
+        self.assertEqual(logits.shape[0], N_TRAIN)
+        # Logits are not a probability simplex.
```
Comment (anthonycaterini) [inline tests/test_inference.py]: what does this comment mean? why wouldn't they be a simplex?

File: tests/test_inference.py [hunk 12]
```diff
@@ -0,0 +1,149 @@
...
+
+    def test_ensemble_predict_proba_permute_flag(self):
+        for permute in (True, False):
+            p = self.model.ensemble_predict_proba(
+                self.X, n_ensembles=2, context_size=2048, permute_classes=permute, seed=1
+            )
+            self.assertEqual(p.shape, (N_TRAIN, 3))
+            np.testing.assert_allclose(p.sum(axis=1), 1.0, atol=1e-4)
+
+    def test_class_perm(self):
```
Comment (anthonycaterini) [inline tests/test_inference.py:59]: should we test that doing the class perm like this is equivalent to doing the normal prediction and then permuting after?

File: tests/test_inference.py [hunk 13]
```diff
@@ -0,0 +1,149 @@
...
+
+class TestLargeClass(unittest.TestCase):
+    """num_classes > max_num_classes triggers the digit-decomposition path."""
+
+    @classmethod
+    def setUpClass(cls):
+        seed_everything(1)
+        n_classes = 18  # > max_num_classes (16) for the v1.2 weights
+        cls.n_classes = n_classes
+        cls.X = np.random.normal(size=(n_classes * 10, 5)).astype(np.float32)
```
Comment (anthonycaterini) [inline tests/test_inference.py]: this `10` can be defined to ensure consistency between this row and the one below

File: tests/test_reg.py [hunk 14]
```diff
@@ -0,0 +1,50 @@
...
+        """Setup dataset and the v1.2 model (default weights), then gather predictions."""
+        seed_everything(42)
+        X, y = fetch_california_housing(return_X_y=True)
+        cls.X_train, cls.X_test, cls.y_train, cls.y_test = train_test_split(X[:4096], y[:4096], test_size=0.33, random_state=42)
+
+        cls.base_model = TabDPTRegressor(device=DEVICE)
+        cls.base_model.fit(cls.X_train, cls.y_train)
+
+        cls.y_train_pred = cls.base_model.predict(cls.X_train, n_ensembles=8, context_size=None, seed=42)
+        cls.y_test_pred = cls.base_model.predict(cls.X_test, n_ensembles=8, context_size=None, seed=42)
```
Comment (anthonycaterini) [inline tests/test_reg.py]: suggest defining `4096`, `n_ensembles`, `context_size`, and `seed` at the top of this function for consistency in these two lines

File: paper_evaluation.py [hunk 15]
```diff
@@ -163,23 +207,42 @@
...
+    ]
+    for suite in ["cc18", "ctr23", "tabarena"]:
+        sub = df[df["suite"] == suite]
+        if sub.empty:
+            continue
+        lines.append(f"[{suite}] datasets={len(sub)}  IQM:")
+        for metric in ["acc", "auc", "corr", "r2"]:
+            val = robust_iqm(sub[metric])
+            if val is not None:
+                lines.append(f"    {metric}: {val:.4f}")
```
Comment (anthonycaterini) [inline paper_evaluation.py]: why not use `\t` instead of the spaces

File: pyproject.toml [hunk 16]
```diff
@@ -43,9 +43,19 @@ Repository = "https://github.com/layer6ai-labs/TabDPT-inference"
 Paper = "https://arxiv.org/abs/2410.18164"
 Model = "https://huggingface.co/Layer6/TabDPT"
 
+[tool.pytest.ini_options]
+testpaths = ["tests"]
```
Comment (anthonycaterini) [inline pyproject.toml:47]: **Antigravity comment**

Running pytest directly from the repository root without first doing an editable install (pip install -e .) fails because neither src/ nor tests/ is in sys.path. Furthermore, the test modules use top-level imports like from device_utils import pick_device.

Recommendation: Add pythonpath = ["src", "tests"] to [tool.pytest.ini_options]

File: paper_evaluation.py
Comment (anthonycaterini) [inline paper_evaluation.py]: **Antigravity comment**: 

Setting os.environ["CUDA_VISIBLE_DEVICES"] after PyTorch has already been loaded is unreliable and often ignored by the CUDA runtime. Recommendation: Explicitly pass the device ordinal to PyTorch: 

```python
if args.use_cpu:
    device = "cpu"
else:
    device = f"cuda:{args.gpu_to_use}"
```

File: tests/test_estimator.py [hunk 17]
```diff
@@ -0,0 +1,84 @@
...
+        ("power", PowerTransformer),
+        ("quantile-uniform", QuantileTransformer),
+        ("quantile-normal", QuantileTransformer),
+        ("log1p", Log1pScaler),
+        (None, type(None)),
+    ],
+)
+def test_normalizer_selects_scaler(normalizer, scaler_cls):
+    est = build(normalizer=normalizer)
+    assert isinstance(est.scaler, scaler_cls)
```
Comment (anthonycaterini) [inline tests/test_estimator.py:67]: **Antigravity comment**

Because `build()` instantiates `TabDPTEstimator`, every iteration re-reads the full safetensors weights file and reconstructs the PyTorch model from scratch. This causes 8 full model loads just to assert scaler class mapping (plus 3 more in `test_invalid_*_raises`).

Recommendation: Either mock model weight loading for scaler mapping checks, or cache a shared base model and test the scaler initialization logic independently.

File: tests/test_reg.py [hunk 18]
```diff
@@ -0,0 +1,50 @@
...
+
+    def test_accuracy(self):
+        """Train prediction should be near-perfect; v1.2 test R2 should be at least v1.1's."""
+        train_r2 = r2_score(self.y_train, self.y_train_pred)
+        test_r2 = r2_score(self.y_test, self.y_test_pred)
+        print(f"\n[reg] v1.2 test R2 {test_r2:.4f} vs v1.1 {V1_REG_R2:.4f} (delta {test_r2 - V1_REG_R2:+.4f})")
+
+        self.assertGreaterEqual(
+            train_r2, TRAIN_MIN_R2 - TOLERANCE,
+            msg=f"train R2 {train_r2:.4f} below expectation {TRAIN_MIN_R2}",
```
Comment (anthonycaterini) [inline tests/test_reg.py]: `msg` is slightly wrong. Should say `f"train R2 {train_r2:.4f} is over {TOLERANCE} below expectation {TRAIN_MIN_R2}"`

File: paper_evaluation.py [hunk 19]
```diff
@@ -30,6 +40,14 @@
     parser.add_argument("--use-cpu", action="store_true", help="If true, use CPU for evalutation")
```
Comment (anthonycaterini) [inline paper_evaluation.py:40]: typo here (`evaluation` -> `evalutation`)

File: paper_evaluation.py
Comment (anthonycaterini) [inline paper_evaluation.py]: `context_size` uses underscore, everything else uses dash (`n-ensembles`, `batch-size`, etc.)

**Review by anthonycaterini — CHANGES_REQUESTED** [review]
Pretty much everything is minor. Had antigravity also check. Additional comments:

1. I think `README.md` should be further updated to give instructions for running tests
2. I have not run the tests myself yet because I don't have GPU. @j201 have you run them?

File: tests/device_utils.py (hunk 2 above)
Comment (j201) [reply tests/device_utils.py:7]: Why? It makes it harder to see what the defaults are when reading the functions or using function signature previews in IDEs. The constants are only used once and in a way that makes their meaning clear.

File: tests/test_cls.py (hunk 6 above)
Comment (j201) [reply tests/test_cls.py]: I think it'd actually be better to just use the default config, except for seeds, since future changes might necessitate different defaults for good predictive performance (this already varies from v1.1 in the context_size setting). It's more relevant to check for regressions in default predictive performance than on configs that might no longer apply. Will change

File: tests/test_cls.py (hunk 7 above)
Comment (j201) [reply tests/test_cls.py:32]: I don't think it's necessary, although we currently get a perfect score and it would be surprising if we didn't given the current way the model works, so it could be useful to keep as a regression test. If we don't have it in the future, the test can be removed.

File: tests/test_cls.py (hunk 8 above)
Comment (j201) [reply tests/test_cls.py]: I think the intent here is to avoid floating point precision issues, but since it's an accuracy score and the dataset isn't that big, the intent is to check if it's 1.0. Will change to `assertAlmostEqual`.

File: tests/test_estimator.py (hunk 9 above)
Comment (j201) [reply tests/test_estimator.py:13]: It's used on line 47

File: tests/test_estimator.py (hunk 17 above)
Comment (j201) [reply tests/test_estimator.py:67]: Can't cache a shared base model because these are testing constructor logic. Perhaps mocking could be done but I'd leave that as a future optimization for now.

File: tests/test_inference.py (hunk 11 above)
Comment (j201) [reply tests/test_inference.py]: I think the intent is to say "Generally, rows shouldn't all sum to one - if they do, we're probably returning probabilities instead of logits.", will change to that.

File: pyproject.toml (hunk 16 above)
Comment (j201) [reply pyproject.toml:47]: Our README instructions say to do an editable install if you're using pip, also this works fine for me with `uv sync`

File: tests/test_inference.py (hunk 10 above)
Comment (j201) [reply tests/test_inference.py]: Changed

File: tests/test_inference.py (hunk 12 above)
Comment (j201) [reply tests/test_inference.py:59]: Added

File: tests/test_inference.py (hunk 13 above)
Comment (j201) [reply tests/test_inference.py]: Changed

File: tests/test_reg.py (hunk 14 above)
Comment (j201) [reply tests/test_reg.py]: Same comment as for the classifier, I think it's better to just go with the defaults

File: tests/test_reg.py (hunk 18 above)
Comment (j201) [reply tests/test_reg.py]: Changed

File: paper_evaluation.py
Comment (j201) [reply paper_evaluation.py]: Changed

File: paper_evaluation.py (hunk 19 above)
Comment (j201) [reply paper_evaluation.py:40]: Fixed

File: paper_evaluation.py
Comment (j201) [reply paper_evaluation.py]: Changed

File: examples/cls_example.py
Comment (j201) [reply examples/cls_example.py:1]: Changed

**Comment by j201** [comment]
Replied to comments, also confirmed that tests all pass as of now

File: tests/device_utils.py (hunk 2 above)
Comment (anthonycaterini) [reply tests/device_utils.py:7]: true, fair enough

File: tests/test_cls.py (hunk 6 above)
Comment (anthonycaterini) [reply tests/test_cls.py]: ok thanks, makes sense

**Review by anthonycaterini — APPROVED** [review]
